### PR TITLE
support NamedTuples

### PR DESCRIPTION
**Pull request by aplavin**

### CODE REVIEW

File: src/InvertedIndices.jl
Comment (bkamins) [inline src/InvertedIndices.jl]: following on the discussion from the previous PR opened by @ararslan  - what is the status in Compat.jl? Can these be also safely used pre Julia 1.7 or not?

Comment (aplavin) [reply src/InvertedIndices.jl]: Don't see this namedtuple indexing after a cursory look at Compat.jl. Never used or looked at that package before, so cannot say much more.

Comment (bkamins) [reply src/InvertedIndices.jl]: What I mean by my question is that, if I understood @ararslan correctly, we should first check the status of Compat.jl before merging this PR.
InvertedIndices.jl currently guarantees compatibility with Julia 1.1 or later, so we should make sure all codes using InvertedIndices.jl will work under these different versions of Julia (I understand that the indexing without `Not` was introduced in Julia 1.7, but at least we should decide before merging if we want to accept the fact that this package does not strictly support all Julia versions since 1.1 in the same way).

Comment (aplavin) [reply src/InvertedIndices.jl]: How is the current status of Compat.jl relevant here? InvertedIndices doesn't depend on Compat anyway, and `using InvertedIndices` won't bring the new (1.7) namedtuple indexing.
Whether these `Not` indexing methods are defined on 1.6- versions or not, the result would be effectively the same: a `MethodError`.

Comment (aplavin) [reply src/InvertedIndices.jl]: As for implementations based on @ararslan's suggestions, 
```
@inline Base.getindex(nt::NamedTuple, I::InvertedIndex{Symbol}) = (; (k => nt[k] for k in keys(nt) if k != I.skip)...)
@inline Base.getindex(nt::NamedTuple, I::InvertedIndex{NTuple{N, Symbol}}) where {N} = (; (k => nt[k] for k in keys(nt) if k ∉ I.skip)...)
```
they fail the `@inferred` tests.

Comment (bkamins) [reply src/InvertedIndices.jl]: > How is the current status of Compat.jl relevant here?

We would start depending on Compat.jl.

> using `InvertedIndices` won't bring the new (1.7) namedtuple indexing.

We could discuss if we want to bring it or not.

---

My point is that if the user writes `nt[Not((:a, :b))]` it should work under all supported Julia versions (i.e. anything starting from 1.1) - this is the contract we currently have in InvertedIndices.jl. If we were to change it we need to discuss this.

The reason why I am so strict is that InvertedIndices.jl is used by 609 packages directly or indirectly so we cannot lightly make changes in this package (even if they are "borderline" decisions). We have to have a detailed consideration before we make any change here.

Comment (aplavin) [reply src/InvertedIndices.jl]: It's hard to say something definite for me here, as I wasn't aware of the contract:
> it should work under all supported Julia versions (i.e. anything starting from 1.1) - this is the contract we currently have in InvertedIndices.jl

Depending on how far this contract goes, it could be possible to use the current implementation in 1.7+, and a type-unstable implementation (see above) in earlier versions. Still, some tests would only work on 1.7+: the `@inferred` ones.

Comment (aplavin) [reply src/InvertedIndices.jl]: Having read the docs, I thought that `x[Not(i)]` should effectively give the complement to `x[i]`. This holds with the current implementation: when `nt[(:a, :b)]` works, `nt[Not((:a, :b))]` does as well.

Comment (bkamins) [reply src/InvertedIndices.jl]: That is why I call this case "borderline".

I have also checked the docstring.

If we want to merge this PR (which I personally think we want eventually when we sort out what to do) we should update the docstring to describe the behavior implemented.

Maybe the easiest solution is to document in a separate section of the docstring that this feature is supported only for Julia 1.7.0 and we will be fine. In the new part of the docstring also can you please add that in this case indexing is with names not numbers and that bounds (in this case if the names are present in the queried `NamedTuple`) are not checked.

Comment (aplavin) [reply src/InvertedIndices.jl]: What do you think of the current implementation? All tests pass on 1.6 and 1.7, don't have earlier versions installed.

Comment (bkamins) [reply src/InvertedIndices.jl]: Looks good. @ararslan - OK with you?

But can you please add the description to the docstring as we have a bit different rules here than for normal `Not` anyway. Thank you!

File: src/InvertedIndices.jl
Comment (bkamins) [inline src/InvertedIndices.jl]:
```suggestion
When indexing into a `NamedTuple` with `InvertedIndex` it accepts either a `Symbol`
or a tuple of `Symbol`s and selects the fields of the `NamedTuple` that
are not in `idx`. For `NamedTuple` bounds checking of `idx` is not performed:
```

**Review by bkamins — APPROVED** [review]

**Comment by nalimilan** [comment]
I'm not convinced that we should silently skip names that are not in the `NamedTuple`. I saw the discussion at https://github.com/JuliaData/InvertedIndices.jl/pull/28#discussion_r761866555, but the behavior of `Base.structdiff` doesn't sound like a strong argument to me, given that it's not exported and that this behavior isn't even documented. I admit that `delete!` allows non-existent keys contrary to `deleteat!`, but I'm not sure it's enough to consider that it's a general policy (`delete!` doesn't work for `NamedTuple` BTW).

I see several counter arguments:
- We do not allow skipping integer indices that do not exist, even if that would have been a possible choice. Why should we do something different for symbol indices?
- In DataFrames we throw an error when `Not` is passed with a non-existent column name.
- If we throw an error, we can change our minds later and allow non-existent names. If we allow them, we can't change that without breaking code.

**Comment by aplavin** [comment]
I personally prefer the stricter behavior too, and wish functions like `Base.delete!` also threw exceptions on non-existing keys.
In this PR I originally followed the Base behavior for consistency: that is, removing a nonexisting element from "key-indexed" collections doesn't throw, unlike "int-indexed" ones.

Not sure what is the decision status on "strictness" for keyed `Not` indexing now. Please see the updated code that throws the same exception as `nt[:nonexistent]` access.

**Comment by bkamins** [comment]
I guess this is what @nalimilan wanted

**Comment by nalimilan** [comment]
OK, if we all think being strict is best, it's more logical to go with that (at least for now).

Any reason not to allow using integer indices too? That's already supported for tuples.

File: test/runtests.jl
Comment (nalimilan) [inline test/runtests.jl]:
```suggestion
@testset "named tuples" begin
```

**Comment by aplavin** [comment]
> Any reason not to allow using integer indices too? That's already supported for tuples.

Are they?
```
julia> (1, 2, 3)[Not(1)]
ERROR: MethodError: no method matching getindex(::Tuple{Int64, Int64, Int64}, ::InvertedIndex{Int64})
```

**Comment by nalimilan** [comment]
Woops, I confused with code that allows indexing arrays with a tuple.

Would you mind adding support for indexing both tuples and named tuple with `Not` wrapping integers? It sounds weird to support only named tuples with `Not` wrapping symbols.

**Comment by bkamins** [comment]
@nalimilan - indeed it is weird. However, this should ba added only for `Tuple`, and not for `NamedTuple`. The problem is that `NamedTuple` defines the following methods only:
```
julia> methods(getindex, (NamedTuple, Any))
# 4 methods for generic function "getindex":
[1] getindex(t::NamedTuple, i::Int64) in Base at namedtuple.jl:126
[2] getindex(t::NamedTuple, i::Symbol) in Base at namedtuple.jl:127
[3] getindex(t::NamedTuple, idxs::Tuple{Vararg{Symbol}}) in Base at namedtuple.jl:128
[4] getindex(t::NamedTuple, idxs::AbstractVector{Symbol}) in Base at namedtuple.jl:129
```

so negation of an integer index does not make sense (it is not supported). I would say that in general the design of indexing for `NamedTuple` seems a bit strange, but we have to live with it.

While for `Tuple` we have:
```
julia> methods(getindex, (Tuple, Any))
# 7 methods for generic function "getindex":
[1] getindex(t::Tuple, i::Int64) in Base at tuple.jl:29
[2] getindex(t::Tuple, i::Real) in Base at tuple.jl:30
[3] getindex(t::Tuple, c::Colon) in Base at tuple.jl:33
[4] getindex(t::Tuple, r::AbstractUnitRange) in Base at range.jl:369
[5] getindex(t::Tuple, b::AbstractVector{Bool}) in Base at tuple.jl:32
[6] getindex(t::Tuple, r::AbstractVector) in Base at tuple.jl:31
[7] getindex(t::Tuple, i::CartesianIndex{1}) in Base at multidimensional.jl:832
```
However, I am OK to make `Not` for `Tuple` a separate PR - as @aplavin prefers.

**Review by bkamins — CHANGES_REQUESTED** [review]
@aplavin I have just noticed that we should also add support for `Not` on `AbstractVector{Symbol}` for `NamedTuple`.

**Comment by aplavin** [comment]
bump...
anything else left to implement for namedtuples?

File: src/InvertedIndices.jl
Comment (bkamins) [inline src/InvertedIndices.jl]:
```suggestion
When indexing into a `NamedTuple` with `InvertedIndex` it accepts either a `Symbol`,
a vector of `Symbol`s, or a tuple of `Symbol`s and selects the fields of the `NamedTuple` that
```

File: src/InvertedIndices.jl
Comment (bkamins) [inline src/InvertedIndices.jl]: 1. can you please refactor the code to make lines be narrower (preferably below 92 characters)
2. Instead of `error` we should `throw` a more specific error type here I think. I know that in `NamedTuple` implementation `ErrorException` is thrown, but it seems not a good choice to me. @nalimilan - do you think `ArgumentError` is better? I am also OK with leaving `ErrorException`.

Comment (nalimilan) [reply src/InvertedIndices.jl]: I agree `ErrorException` isn't the best choice, but given that this is what Base throws it makes sense to do the same (until it's changed there).

File: src/InvertedIndices.jl
Comment (nalimilan) [inline src/InvertedIndices.jl]:
```suggestion
When indexing into a `NamedTuple`, the `InvertedIndex` can wrap either a `Symbol`,
```

File: src/InvertedIndices.jl
Comment (bkamins) [inline src/InvertedIndices.jl:173]:
```suggestion
    allunique(I.skip) && I.skip ⊆ keys(nt) ? Base.structdiff(nt, NamedTuple{I.skip}) :
```
@aplavin + @nalimilan - one last thing to discuss (thank you for your patience and sorry for such a long process but I want to carefully consider all cases).

I propose to add the `allunique` test as this is the rule that normal `NamedTuple` indexing enforces:
```
julia> nt = (a=1,b=2)
(a = 1, b = 2)

julia> nt[(:a, :a)]
ERROR: duplicate field name in NamedTuple: "a" is not unique
```
For vectors uniqueness is not enforced in `Not`:
```
julia> v = 1:3
1:3

julia> v[Not([1,1])]
2-element Vector{Int64}:
 2
 3
```
but AFAICT the reason is that duplicates are allowed with vector indexing:
```
julia> v[[1,1]]
2-element Vector{Int64}:
 1
 1
```
In DataFrames.jl we do not allow duplicates with `Not` on columns:
```
julia> df = DataFrame(rand(3,4), :auto);

julia> df[:, Not([:a, :a])]
ERROR: ArgumentError: Elements of [:a, :a] must be unique
```
(this is not a definitive reference, but it is easier to later allow things as turning error to non-error is non-breaking)

If we agree to this tests would need to be updated.

Thank you!

Comment (bkamins) [reply src/InvertedIndices.jl:173]: ah - and if we agree the error message would need to be updated.

Comment (nalimilan) [reply src/InvertedIndices.jl:173]: Makes sense, but the error message should probably be different, otherwise it would be misleading. I also hope the performance impact isn't too large.

Comment (aplavin) [reply src/InvertedIndices.jl:173]: All these extra checks make the functionality less and less useful due to significantly worse performance:
```
julia> nt = (a=1, b=2., c=:def)

# baseline
julia> @btime Base.structdiff($nt, NamedTuple{(:a, :b)})
  1.587 ns (0 allocations: 0 bytes)

# without any checks: my original PR
julia> @btime $nt[Not((:a, :b))]
  1.540 ns (0 allocations: 0 bytes)

# current: subset check
julia> @btime $nt[Not((:a, :b))]
  3.273 ns (0 allocations: 0 bytes)

# with allunique()
julia> @btime $nt[Not((:a, :b))]
  103.062 ns (4 allocations: 384 bytes)
```
I can hardly see myself using this `Not` indexing with two orders of magnitude overhead...

Comment (bkamins) [reply src/InvertedIndices.jl:173]: Indeed even the constructor in some cases drops this check:
```
julia> (;:a=>1,:a=>2)
(a = 2,)
```
(which is unfortunate I but probably unavoidable)

**Review by bkamins — APPROVED** [review]

**Comment by nalimilan** [comment]
Thanks!

**Pull request merged**
